### DISCUSSION TITLE
[Feature] Transcode setting of average bitrate

### DISCUSSION THREAD

**Discussion by Goodwu** [immich-app/immich#10567] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

![image](https://github.com/immich-app/immich/assets/348005/176e7a69-e936-4f26-87ae-47326181abfc)
The transcoding output quality is very low on J3455 with HW encoding on.
The average bitrate is lower then 500kb/s. 
There's no effect by setting max bitrate to higher value. But set averate bitrate to higher value by command line "-b:v" have effect.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by Goodwu** [comment]
Is this a bug?
I look into the code and I see some other encoders like NV and VAAPI have -b:v argument but Qsv does not.

**Reply by bo0tzz** [reply]
cc @mertalev ?

**Reply by mertalev** [reply]
`b:v` is for target bitrate, not max bitrate. The limit only applies when the encoder would otherwise use a higher bitrate for a scene. You can compare the bitrate without that limit by removing the `maxrate` flag (or setting it to 0 in immich).

**Reply by mertalev** [reply]
For general quality improvement, you can try setting the slowest preset. If it's still not good enough, you can also lower CRF.

**Reply by Goodwu** [reply]
![image](https://github.com/immich-app/immich/assets/348005/2f05ac34-3a86-46f5-b030-ac39ee23729a)
Maybe there's something wrong with ffmpeg. It doesn't reach the request quality and bitrate.
The command line in my test:
`/usr/bin/ffmpeg -hwaccel qsv -hwaccel_output_format qsv -async_depth 4 -threads 1 -i /media/photos/小区相关/2014物业投标/00001.MTS -y -c:v h264_qsv -c:a aac -mov lags faststart -fps_mode passthrough -map 0:0 -map 0:1 -bf 7 -refs 5 -g 256 -v verbose -vf deinterlace_qsv,scale_qsv=-1:720:async_depth=4:mode=hq:format=nv12 -preset 1 -global_quality:v 2 -maxrate 4000k -bufsize 8000k test.mp4`

**Reply by Goodwu** [reply]
The quality of output video is still very horrible even I set the quality to 2 and preset to veryslow.

**Reply by suurt8ll** [reply]
I'm experiencing the same issue with h264_qsv. I discovered that removing `-maxrate` and adding `-b:v 0` makes ffmpeg behave as I (and likely @Goodwu) intend.

@Goodwu, please try running this modified version of your command:

`/usr/bin/ffmpeg -hwaccel qsv -hwaccel_output_format qsv -async_depth 4 -threads 1 -i /media/photos/小区相关/2014物业投标/00001.MTS -y -c:v h264_qsv -c:a aac -mov lags faststart -fps_mode passthrough -map 0:0 -map 0:1 -bf 7 -refs 5 -g 256 -v verbose -vf deinterlace_qsv,scale_qsv=-1:720:async_depth=4:mode=hq:format=nv12 -preset 1 -global_quality:v 2 -b:v 0 -bufsize 8000k test.mp4`

If I set `-b:v 4500k` and keep `-maxrate` removed, I get this result (`-global_quality:v 15`):

![image](https://github.com/user-attachments/assets/5ba57e26-6a3c-4abc-8048-abaf99dd2853)

Perhaps qsv considers `-b:v` as the maximum bitrate limit instead of `-maxrate`? I will investigate this further.

**Reply by mertalev** [reply]
Do the logs specify which rate control method it's using? You might get more info setting `-v debug`. The [Intel docs](https://github.com/intel/media-delivery/blob/master/doc/quality.rst#h-264-avc) say that `-b:v` + `-maxrate` where the latter is higher than the former triggers VBR mode, but I also see the [FFmpeg doc](https://trac.ffmpeg.org/wiki/Hardware/QuickSync)'s example for VBR just has `-b:v`.

**Reply by suurt8ll** [reply]
`ffmpeg -stats -report -hwaccel qsv -hwaccel_output_format qsv -async_depth 4 -noautorotate -threads 1 -i input.mp4 -y -c:v h264_qsv -c:a copy -movflags faststart -fps_mode passthrough -map 0:0 -map 0:1 -bf 7 -refs 5 -g 256 -v debug -vf scale_qsv=-1:720:async_depth=4:mode=hq:format=nv12 -preset 4 -global_quality:v 23 -maxrate 4500k output.mp4`
^ This command is what Immich will currently run if the user sets maxrate in the frontend and uses hardware for transcoding.
Logs say QVBR gets used in this case, bitrate is stuck at 1k:
![image](https://github.com/user-attachments/assets/df288b01-2761-4a7f-b841-cc5bcf15040c)

After removing `-maxrate 4500k`:
![image](https://github.com/user-attachments/assets/37533787-6a77-43d2-9952-8785a95df388)
Now it uses ICQ, and bitrate is no longer stuck at 1k.

After removing `-global_quality:v 23` and using `-b:v 2000k -maxrate:v 4500k`:
![image](https://github.com/user-attachments/assets/c55d2f64-7738-428c-8deb-35946eeef66a)
This triggers VBR like the Intel docs say.

**Reply by suurt8ll** [reply]
I think qsv passes `-b:v 1k` always by default if it's not set in the arguments for some reason. `-b:v 1k -maxrate 4500k -global_quality:v 23` all together trigger QVBR not ICQ.

**Reply by suurt8ll** [reply]
OK in the meantime I'll use this workaround for the 1k limit issue:
1. Set _Maximum bitrate_ to 0 in System Settings
2. Choose _Constant rate factor_
3. In _Transcode policy_ select _All videos_

This works, but effectively loses user the ability to transcode videos based on max bitrate policy. _All videos_ policy will waste space but in my case it's not a huge problem because I do not store that much videos.

If you want me to run other commands on my hardware for debugging I'm happy to help out! Thanks for your work!

**Reply by mertalev** [reply]
What rate control mode does it use if you set `-b:v 0 -maxrate 4500k -global_quality:v 23`? Maybe we just need to make sure we add `-b:v 0` if we're adding `-maxrate`.

**Reply by suurt8ll** [reply]
> [h264_qsv @ 0x563132db0f40] Using input frames context (format qsv) with h264_qsv encoder.
> [h264_qsv @ 0x563132db0f40] Encoder: input is video memory surface
> [h264_qsv @ 0x563132db0f40] Use Intel(R) oneVPL to create MFX session with the specified MFX loader
> [h264_qsv @ 0x563132db0f40] Using the constant quality with VBR algorithm (QVBR) ratecontrol method
> [h264_qsv @ 0x563132db0f40] Error initializing the encoder: invalid video parameters (-15)
> [vost#0:0/h264_qsv @ 0x563132d518c0] Error initializing output stream: Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height
> [AVIOContext @ 0x563132dd7c00] Statistics: 0 bytes written, 0 seeks, 0 writeouts
> Terminating demuxer thread 0
> [AVIOContext @ 0x563132d52c40] Statistics: 1386984 bytes read, 2 seeks
> Conversion failed!

- `-b:v 0 -maxrate 4500k -global_quality:v 23` chooses QVBR but fails, see provided log above.
- `-b:v 0 -maxrate 4500k` chooses VBR but also fails in the same way.
- `-b:v 0` chooses CBR but same failure

`-b:v 0 -global_quality:v 23` is the only combo where bitrate can be 0 and command does not fail (this uses ICQ). In fact `-b:v` seems to have 0 importance in this configuration. It does not change the resulting bitrate, whatever number i pick.

Only when `-maxrate` is introduced QSV picks QVBR rate control mode for some reason and starts behaving weirdly.

**Reply by mertalev** [reply]
Thanks for testing! It's a shame that it doesn't work. We might have to handle it like [VAAPI](https://github.com/immich-app/immich/blob/c9c56ac6007abb1d599fc202d47c44cc2ddf69fb/server/src/utils/media.ts#L834) and switch to VBR when the max rate is set.
